File: 0318/SeoyeonHong/1654 랜선 자르기.py
# 만들 수 있는 최대 랜선의 길이
import sys

input = sys.stdin.readline
K, N = map(int, input().split())
cables = []
for _ in range(K):
    cables.append(int(input()))
cables.sort()

s, l = 1, cables[K-1]
while s <= l:
    m = (s + l) // 2
    count = 0
    for cable in cables:
        count += cable // m
    if count < N:
        l = m - 1
    else:
        s = m + 1

print(l)

File: 0318/SoohyunKim/1654_랜선자르기.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const [K, N] = input.shift().split(' ').map(Number);
const lan = input.map(x => Number(x)).sort();
let answer = 0;

// 랜선 길이 최소값과 최대값
let start = 1;
let end = Math.max(...lan);

// 이분 탐색
while (start <= end) {
  const mid = Math.floor((start + end) / 2);
  let count = 0;

  for (let i = 0; i < K; i++) {
    count += Math.floor(lan[i] / mid);
  }

  if (count < N) end = mid - 1; // 길이 줄임
  else {
    answer = mid;
    start = mid + 1; // 길이 늘림
  }
}

console.log(answer);